File: SPOJ/PIGBANK.py
for _ in xrange(input()):
	e,f = map(int,raw_input().split())
	wt = f-e
	type_coins = input()
	val_list = []
	coin_wt = []
	for _ in xrange(type_coins):
		p,w = map(int, raw_input().split())
		val_list.append(p)
		coin_wt.append(w)
	dp = [10**10 for i in xrange(wt+1)]
	# dp[0] = 0
	for j in xrange(type_coins):
		for i in xrange(1,wt+1):
			if coin_wt[j] == i:
				dp[i] = min(dp[i],val_list[j])
			if coin_wt[j] < i:
				if dp[i-coin_wt[j]] != 10**10:
					dp[i] = min(dp[i],dp[i-coin_wt[j]] + val_list[j])
	if dp[wt] != 10**10:
		print "The minimum amount of money in the piggy-bank is %d." %dp[wt]
	else:
		print "This is impossible." 

File: SPOJ/PIGBANK.cpp
#include <bits/stdc++>
#define max1 2147483647
using namespace std;

long int min(long int a,long int b){
	if a < b{
		return a;
	}
	return b;
}

int main(){
	unorderedmap<int,int> mymap;
	int empty,full,coins,weight,val;
	cin >> empty >> full;
	vector<long int> dp(full-empty + 1,max1);
	cin >> coins;
	int wt[coins];
	for(int i=0;i<coins;i++){
		cin >> val >> weight;
		mymap[weight] = val;
		wt[i] = weight;
	}
	dp[0] = 0;
	for(int i = 1;i < full-empty + 1;i++){
		for(int j = 0;j < coins;j++){
			if (wt[j] <= i){
				if dp[i-wt[j]] != max1{
					dp[i] = min(dp[i-wt[j]] + mymap[wt[j]],dp[i]);
				}
			}
		}
	}
	if (dp[full-empty] == max1){
		cout << "The minimum amount of money in the piggy-bank is " << dp[full-empty] << "." << endl;
	}
	else{
		cout << "This is impossible." << endl;
	}
	return 0;
}